### PR TITLE
RSDK-2172 Add Unit Tests for SLAM Interface Functions

### PR DESCRIPTION
**Pull request by jeremyrhyde**
This PR adds unit tests of the SLAM service interface functions in viam-cartographer and uses the new injectable SLAM client to handle the GRPC connection.

Testing was done in concert with the rdk PR linked below (the replace function in go.mod was used to reference this local branch of rdk). Github CLI is expected to fail until the associated PR is push. Once pushed, the `go.mod` will be updated to the new commit of rdk, allowing this to pass.

JIRA Ticket: https://viam.atlassian.net/browse/RSDK-2172
Associated PR: https://github.com/viamrobotics/rdk/pull/2206

### CODE REVIEW

File: viam-cartographer_internal_test.go [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+		return nil, err
+	}
+
+	resp := &v1.GetInternalStateResponse{
+		InternalStateChunk: clientBuffer[:n],
+	}
+	return resp, nil
+}
+
+// func makeGetPositionResponse(pose *spatialmath.Pose, componentReference *string) (*v1.GetPositionResponse, error) {
```
Comment (nicksanford) [inline viam-cartographer_internal_test.go:78]: Why is this here?

File: viam-cartographer_internal_test.go [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+	chunkSizeInternalState     int
+	inputComponentRef          string
+	expectedPointCloudMapBytes []byte
+	expectedInternalStateBytes []byte
+	inputQuat                  map[string]interface{}
+	inputPose                  commonv1.Pose
+)
+
+type pointCloudClientMock struct {
+	grpc.ClientStream
```
Comment (nicksanford) [inline viam-cartographer_internal_test.go:33]: why does this need to inherit from `grpc.ClientStream`?

File: viam-cartographer_internal_test.go [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+}
+
+func TestGetPositionEndpoint(t *testing.T) {
+	svc := &cartographerService{}
+	mockSLAMClient := &inject.SLAMServiceClient{}
+	svc.clientAlgo = mockSLAMClient
+
+	// Add successful GetPosition client
+	mockSLAMClient.GetPositionFunc = func(ctx context.Context, in *v1.GetPositionRequest, opts ...grpc.CallOption) (*v1.GetPositionResponse, error) {
+		resp, _ := makeGetPositionResponse(&inputPose, &inputQuat, &inputComponentRef, "quat")
```
Comment (nicksanford) [inline viam-cartographer_internal_test.go:137]: Why are we not checking for errors?

File: viam-cartographer_internal_test.go [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+		inputComponentRef = "componentReference"
+		pose, componentRef, err := svc.GetPosition(context.Background())
+		test.That(t, err.Error(), test.ShouldContainSubstring, "quaternion given, but invalid format detected")
+		test.That(t, pose, test.ShouldBeNil)
+		test.That(t, componentRef, test.ShouldEqual, "")
+	})
+
+	// Add invalid GetPosition client
+	mockSLAMClient.GetPositionFunc = func(ctx context.Context, in *v1.GetPositionRequest, opts ...grpc.CallOption) (*v1.GetPositionResponse, error) {
+		resp, _ := makeGetPositionResponse(&inputPose, &inputQuat, &inputComponentRef, "not_quat")
```
Comment (nicksanford) [inline viam-cartographer_internal_test.go:205]: check for error?

File: viam-cartographer_internal_test.go (hunk 1 above)
Comment (jeremyrhyde) [inline viam-cartographer_internal_test.go:78]: [note] Another option for building the GetPositionResponse but provides less flexibility in the failure tests we can perform

File: viam-cartographer_internal_test.go [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+	})
+
+	// Add failing GetPosition client
+	mockSLAMClient.GetPositionFunc = func(ctx context.Context, in *v1.GetPositionRequest, opts ...grpc.CallOption) (*v1.GetPositionResponse, error) {
+		return nil, errors.New("error in get position")
+	}
+
+	t.Run("error return failure", func(t *testing.T) {
+		pose, componentRef, err := svc.GetPosition(context.Background())
+		test.That(t, err.Error(), test.ShouldContainSubstring, "error in get position")
```
Comment (nicksanford) [inline viam-cartographer_internal_test.go:226]:
```suggestion
		test.That(t, err, test.ShouldBeError)
		test.That(t, err.Error(), test.ShouldContainSubstring, "error in get position")
```

File: viam-cartographer_internal_test.go [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+		resp, _ := makeGetPositionResponse(&inputPose, &inputQuat, &inputComponentRef, "not_quat")
+		return resp, nil
+	}
+
+	t.Run("no quat key found in extra", func(t *testing.T) {
+		inputPose = commonv1.Pose{X: 0, Y: 0, Z: 0, OX: 0, OY: 0, OZ: 1, Theta: 0}
+		inputQuat = map[string]interface{}{"real": 0.0, "imag": 0.0, "jmag": 0.0, "kmag": 0.0}
+		inputComponentRef = "componentReference"
+		pose, componentRef, err := svc.GetPosition(context.Background())
+		test.That(t, err.Error(), test.ShouldContainSubstring, "quaternion not given")
```
Comment (nicksanford) [inline viam-cartographer_internal_test.go:214]:
```suggestion
                test.That(t, err, test.ShouldBeError)
		test.That(t, err.Error(), test.ShouldContainSubstring, "quaternion not given")
```

File: viam-cartographer_internal_test.go [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+		test.That(t, pose, test.ShouldBeNil)
+		test.That(t, componentRef, test.ShouldEqual, "")
+	})
+
+	t.Run("invalid quat map values failure", func(t *testing.T) {
+		inputPose = commonv1.Pose{X: 0, Y: 0, Z: 0, OX: 0, OY: 0, OZ: 1, Theta: 0}
+		inputQuat = map[string]interface{}{"real": "", "imag": 0.0, "jmag": 0.0, "kmag": 0.0}
+		inputComponentRef = "componentReference"
+		pose, componentRef, err := svc.GetPosition(context.Background())
+		test.That(t, err.Error(), test.ShouldContainSubstring, "quaternion given, but invalid format detected")
```
Comment (nicksanford) [inline viam-cartographer_internal_test.go:198]:
```suggestion
                test.That(t, err, test.ShouldBeError)
		test.That(t, err.Error(), test.ShouldContainSubstring, "quaternion given, but invalid format detected")
```

File: viam-cartographer_internal_test.go [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+		test.That(t, pose, test.ShouldResemble, expectedPose)
+		test.That(t, componentRef, test.ShouldEqual, inputComponentRef)
+	})
+
+	t.Run("invalid quat map keys failure", func(t *testing.T) {
+		inputPose = commonv1.Pose{X: 0, Y: 0, Z: 0, OX: 0, OY: 0, OZ: 1, Theta: 0}
+		inputQuat = map[string]interface{}{"r": 1.0, "imag": 0.0, "jmag": 0.0, "kmag": 0.0}
+		inputComponentRef = "componentReference"
+		pose, componentRef, err := svc.GetPosition(context.Background())
+		test.That(t, err.Error(), test.ShouldContainSubstring, "quaternion given, but invalid format detected")
```
Comment (nicksanford) [inline viam-cartographer_internal_test.go:188]:
```suggestion
                 test.That(t, err, test.ShouldBeError)
		test.That(t, err.Error(), test.ShouldContainSubstring, "quaternion given, but invalid format detected")
```

File: viam-cartographer_internal_test.go [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+	})
+
+	// Add failing GetPointCloudMap client
+	mockSLAMClient.GetPointCloudMapFunc = func(ctx context.Context, in *v1.GetPointCloudMapRequest, opts ...grpc.CallOption) (v1.SLAMService_GetPointCloudMapClient, error) {
+		return nil, errors.New("error in get pointcloud map")
+	}
+
+	t.Run("error return failure", func(t *testing.T) {
+		callback, err := svc.GetPointCloudMap(context.Background())
+		test.That(t, err.Error(), test.ShouldContainSubstring, "error in get pointcloud map")
```
Comment (nicksanford) [inline viam-cartographer_internal_test.go:278]:
```suggestion
                test.That(t, err, test.ShouldBeError)
		test.That(t, err.Error(), test.ShouldContainSubstring, "error in get pointcloud map")
```

File: viam-cartographer_internal_test.go [hunk 10]
```diff
@@ -0,0 +1,332 @@
...
+	})
+
+	// Add failing GetInternalState client
+	mockSLAMClient.GetInternalStateFunc = func(ctx context.Context, in *v1.GetInternalStateRequest, opts ...grpc.CallOption) (v1.SLAMService_GetInternalStateClient, error) {
+		return nil, errors.New("error in get internal state")
+	}
+
+	t.Run("error return failure", func(t *testing.T) {
+		callback, err := svc.GetInternalState(context.Background())
+		test.That(t, err.Error(), test.ShouldContainSubstring, "error in get internal state")
```
Comment (nicksanford) [inline viam-cartographer_internal_test.go:329]:
```suggestion
                 test.That(t, err, test.ShouldBeError)
		test.That(t, err.Error(), test.ShouldContainSubstring, "error in get internal state")
```

**Review by nicksanford — CHANGES_REQUESTED** [review]

File: viam-cartographer_internal_test.go (hunk 1 above)
Comment (nicksanford) [reply viam-cartographer_internal_test.go:78]: I think we can remove this option

**Review by nicksanford — APPROVED** [review]

File: viam-cartographer_internal_test.go [hunk 11]
```diff
@@ -0,0 +1,348 @@
...
+		return nil, err
+	}
+
+	resp := &v1.GetInternalStateResponse{
+		InternalStateChunk: clientBuffer[:n],
+	}
+	return resp, nil
+}
+
+func makeGetPositionResponse(
```
Comment (jeremyrhyde) [inline viam-cartographer_internal_test.go:78]: Let me know if we want to add comments here and the makeQuaternionFromProto function below. They seemed self-explanatory to me but can add if desired

Comment (nicksanford) [reply viam-cartographer_internal_test.go:78]: nope I get it

File: viam-cartographer_internal_test.go [hunk 12]
```diff
@@ -0,0 +1,348 @@
...
+)
+
+var (
+	chunkSizePointCloudMap     int
+	chunkSizeInternalState     int
+	inputComponentRef          string
+	expectedPointCloudMapBytes []byte
+	expectedInternalStateBytes []byte
+	inputQuat                  map[string]interface{}
+	inputPose                  commonv1.Pose
```
Comment (kkufieta) [inline viam-cartographer_internal_test.go:29]: [q] Is there any chance that the tests below can be executed in parallel? If so, might it be better to define `inputComponentRef`, `inputQuat`, and `inputPose` in each test separately?

File: viam-cartographer_internal_test.go [hunk 13]
```diff
@@ -0,0 +1,348 @@
...
+			r3.Vector{inputPose.X, inputPose.Y, inputPose.Z},
+			makeQuaternionFromProto(inputQuat),
+		)
+		test.That(t, pose, test.ShouldResemble, expectedPose)
+		test.That(t, componentRef, test.ShouldEqual, inputComponentRef)
+	})
+
+	t.Run("non origin pose success", func(t *testing.T) {
+		inputPose = commonv1.Pose{X: 5, Y: 5, Z: 5, OX: 1, OY: 0, OZ: 1, Theta: 0}
+		inputQuat = map[string]interface{}{"real": 1.0, "imag": 0.0, "jmag": 0.0, "kmag": 0.0}
```
Comment (kkufieta) [inline viam-cartographer_internal_test.go:137]: [q] Why did you change the orientation of the pose but not the quaternion? Why do we use a pose and a quaternion, when the pose already contains the orientation?

File: viam-cartographer_internal_test.go [hunk 14]
```diff
@@ -0,0 +1,348 @@
...
+		Kmag: quat["kmag"].(float64),
+	}
+}
+
+func TestGetPositionEndpoint(t *testing.T) {
+	svc := &cartographerService{}
+	mockSLAMClient := &inject.SLAMServiceClient{}
+	svc.clientAlgo = mockSLAMClient
+
+	// Add successful GetPosition client
```
Comment (kkufieta) [inline viam-cartographer_internal_test.go:111]: [nit] I think you should wrap these tests in another `t.Run` instead of using comments. That should make it easier to debug. 
```suggestion
	t.Run("Successful GetPosition client", func(t *testing.T) {
```

File: viam-cartographer_internal_test.go [hunk 15]
```diff
@@ -0,0 +1,348 @@
...
+		inputQuat = map[string]interface{}{"real": "", "imag": 0.0, "jmag": 0.0, "kmag": 0.0}
+		inputComponentRef = "componentReference"
+		pose, componentRef, err := svc.GetPosition(context.Background())
+		test.That(t, err, test.ShouldBeError)
+		test.That(t, err.Error(), test.ShouldContainSubstring, "quaternion given, but invalid format detected")
+		test.That(t, pose, test.ShouldBeNil)
+		test.That(t, componentRef, test.ShouldEqual, "")
+	})
+
+	// Add invalid GetPosition client
```
Comment (kkufieta) [inline viam-cartographer_internal_test.go:185]: [nit] I think you should wrap these tests in another t.Run instead of using comments. That should make it easier to debug.

```suggestion
	t.Run("Invalid GetPosition client", func(t *testing.T) {
```

File: viam-cartographer_internal_test.go [hunk 16]
```diff
@@ -0,0 +1,348 @@
...
+		inputQuat = map[string]interface{}{"real": 0.0, "imag": 0.0, "jmag": 0.0, "kmag": 0.0}
+		inputComponentRef = "componentReference"
+		pose, componentRef, err := svc.GetPosition(context.Background())
+		test.That(t, err, test.ShouldBeError)
+		test.That(t, err.Error(), test.ShouldContainSubstring, "quaternion not given")
+		test.That(t, pose, test.ShouldBeNil)
+		test.That(t, componentRef, test.ShouldEqual, "")
+	})
+
+	// Add failing GetPosition client
```
Comment (kkufieta) [inline viam-cartographer_internal_test.go:206]: [nit] I think you should wrap these tests in another t.Run instead of using comments. That should make it easier to debug.

Same with the cases below.

**Review by kkufieta — CHANGES_REQUESTED** [review]
Nice work! I have a few nits and questions.

File: viam-cartographer_internal_test.go (hunk 11 above)
Comment (EshaMaharishi) [reply viam-cartographer_internal_test.go:78]: +1 to nick

File: viam-cartographer_internal_test.go [hunk 17]
```diff
@@ -0,0 +1,348 @@
...
+			r3.Vector{inputPose.X, inputPose.Y, inputPose.Z},
+			makeQuaternionFromProto(inputQuat),
+		)
+		test.That(t, pose, test.ShouldResemble, expectedPose)
+		test.That(t, componentRef, test.ShouldEqual, inputComponentRef)
+	})
+
+	t.Run("invalid quat map keys failure", func(t *testing.T) {
+		inputPose = commonv1.Pose{X: 0, Y: 0, Z: 0, OX: 0, OY: 0, OZ: 1, Theta: 0}
+		inputQuat = map[string]interface{}{"r": 1.0, "imag": 0.0, "jmag": 0.0, "kmag": 0.0}
```
Comment (EshaMaharishi) [inline viam-cartographer_internal_test.go:165]: [nit] "r" -> "invalid" to make it easy to see what the invalid field is

File: viam-cartographer_internal_test.go [hunk 18]
```diff
@@ -0,0 +1,348 @@
...
+		pose, componentRef, err := svc.GetPosition(context.Background())
+		test.That(t, err, test.ShouldBeError)
+		test.That(t, err.Error(), test.ShouldContainSubstring, "quaternion given, but invalid format detected")
+		test.That(t, pose, test.ShouldBeNil)
+		test.That(t, componentRef, test.ShouldEqual, "")
+	})
+
+	t.Run("invalid quat map values failure", func(t *testing.T) {
+		inputPose = commonv1.Pose{X: 0, Y: 0, Z: 0, OX: 0, OY: 0, OZ: 1, Theta: 0}
+		inputQuat = map[string]interface{}{"real": "", "imag": 0.0, "jmag": 0.0, "kmag": 0.0}
```
Comment (EshaMaharishi) [inline viam-cartographer_internal_test.go:176]: [nit] "" -> "invalid" to make it easy to see what the invalid value is

File: viam-cartographer_internal_test.go [hunk 19]
```diff
@@ -0,0 +1,348 @@
+// Package viamcartographer implements simultaneous localization and mapping.
+// This is an Experimental package.
```
[the remark this answers is not in the document]
Comment (jeremyrhyde) [reply viam-cartographer_internal_test.go:2]: Thanks for the info!

File: viam-cartographer_internal_test.go [hunk 20]
```diff
@@ -0,0 +1,348 @@
...
+
+	resp := &v1.GetPositionResponse{
+		Pose:               pose,
+		ComponentReference: *componentReference,
+		Extra:              extra,
+	}
+	return resp, nil
+}
+
+func makeQuaternionFromProto(quat map[string]interface{}) spatialmath.Orientation {
```
[the remark this answers is not in the document]
Comment (jeremyrhyde) [reply viam-cartographer_internal_test.go:97]: How does `makeQuaternionFromGenericMap` sound?

File: viam-cartographer_internal_test.go (hunk 14 above)
Comment (jeremyrhyde) [reply viam-cartographer_internal_test.go:111]: I'm personally not in favor of having multi-tier t.Runs for so few tests. Since the largest grouping of tests would 5x I dont think the additional organization is needed or looks as clean in the verbose logs. Below is an example of what the multi-tier might do to GetPosition tests:
```
--- PASS: TestGetPositionEndpoint (0.00s)
    --- PASS: TestGetPositionEndpoint/success (0.00s)
        --- PASS: TestGetPositionEndpoint/success/origin_pose (0.00s)
        --- PASS: TestGetPositionEndpoint/success/non_origin_pose (0.00s)
        --- PASS: TestGetPositionEndpoint/success/empty_component_reference (0.00s)
    --- PASS: TestGetPositionEndpoint/failure (0.00s)
        --- PASS: TestGetPositionEndpoint/failure/invalid_quat_map_keys_failure (0.00s)
        --- PASS: TestGetPositionEndpoint/failure/invalid_quat_map_values_failure (0.00s)
        --- PASS: TestGetPositionEndpoint/failure/no_quat_key_found_in_extra (0.00s)
        --- PASS: TestGetPositionEndpoint/failure/error_return (0.00s)
```
vs.
```
--- PASS: TestGetPositionEndpoint (0.00s)
    --- PASS: TestGetPositionEndpoint/origin_pose_success (0.00s)
    --- PASS: TestGetPositionEndpoint/non_origin_pose_success (0.00s)
    --- PASS: TestGetPositionEndpoint/empty_component_reference_success (0.00s)
    --- PASS: TestGetPositionEndpoint/invalid_quat_map_keys_failure (0.00s)
    --- PASS: TestGetPositionEndpoint/invalid_quat_map_values_failure (0.00s)
    --- PASS: TestGetPositionEndpoint/no_quat_key_found_in_extra (0.00s)
    --- PASS: TestGetPositionEndpoint/error_return_failure (0.00s)
    --- PASS: TestGetPositionEndpoint/undefined_GetPosition_function_in_injected_client (0.00s)
```

**Review by EshaMaharishi — APPROVED** [review]
LGTM!

File: viam-cartographer_internal_test.go (hunk 14 above)
Comment (jeremyrhyde) [reply viam-cartographer_internal_test.go:111]: @EshaMaharishi @nicksanford @kkufieta Let me know though if you like option 1 better!

File: viam-cartographer_internal_test.go (hunk 13 above)
Comment (jeremyrhyde) [reply viam-cartographer_internal_test.go:137]: The pose's orientation is ignored by our GetPosition function. This is because cartographer returns the orientation as a quaternion but our proto message defines it as an `OrientationVector`. Since the rust SDK spatial library is not yet available we are returning the orientation in the extra field to be parsed and used by the `spatialmath` lib in RDK.

Comment (jeremyrhyde) [reply viam-cartographer_internal_test.go:137]: That being said I have updated the test to change the used quaternion and leave the unused oreintationvector in pose constant

File: viam-cartographer_internal_test.go (hunk 14 above)
Comment (EshaMaharishi) [reply viam-cartographer_internal_test.go:103]: I don't feel strongly either way since there are few enough tests!

File: viam-cartographer_internal_test.go [hunk 21]
```diff
@@ -219,6 +223,17 @@ func TestGetPositionEndpoint(t *testing.T) {
...
+	// Add nil GetPosition client
+	mockSLAMClient.GetPositionFunc = nil
+
+	t.Run("undefined GetPosition function in injected client", func(t *testing.T) {
+		pose, componentRef, err := svc.GetPosition(context.Background())
+		test.That(t, err, test.ShouldBeError)
+		test.That(t, err.Error(), test.ShouldContainSubstring, "no GetPositionFunc defined for injected SLAM service client")
+		test.That(t, pose, test.ShouldBeNil)
+		test.That(t, componentRef, test.ShouldEqual, "")
+	})
```
Comment (EshaMaharishi) [inline viam-cartographer_internal_test.go:236]: ah, this test is not needed since it is testing the mock itself :)

same below

same below

**Review by EshaMaharishi — APPROVED** [review]
lgtm mod comment

File: viam-cartographer_internal_test.go (hunk 14 above)
Comment (jeremyrhyde) [reply viam-cartographer_internal_test.go:104]: New breakup:
```
=== RUN   TestGetPositionEndpoint
=== RUN   TestGetPositionEndpoint/successful_client
=== RUN   TestGetPositionEndpoint/successful_client/origin_pose_success
=== RUN   TestGetPositionEndpoint/successful_client/non_origin_pose_success
=== RUN   TestGetPositionEndpoint/successful_client/empty_component_reference_success
=== RUN   TestGetPositionEndpoint/successful_client/invalid_quat_map_keys_failure
=== RUN   TestGetPositionEndpoint/successful_client/invalid_quat_map_values_failure
=== RUN   TestGetPositionEndpoint/successful_client/empty_quat_map_failure
=== RUN   TestGetPositionEndpoint/invalid_client_no_quat_key_found_in_extra_failure
=== RUN   TestGetPositionEndpoint/invalid_client_error_return_failure
=== RUN   TestGetPositionEndpoint/nil_client_failure
--- PASS: TestGetPositionEndpoint (0.00s)
    --- PASS: TestGetPositionEndpoint/successful_client (0.00s)
        --- PASS: TestGetPositionEndpoint/successful_client/origin_pose_success (0.00s)
        --- PASS: TestGetPositionEndpoint/successful_client/non_origin_pose_success (0.00s)
        --- PASS: TestGetPositionEndpoint/successful_client/empty_component_reference_success (0.00s)
        --- PASS: TestGetPositionEndpoint/successful_client/invalid_quat_map_keys_failure (0.00s)
        --- PASS: TestGetPositionEndpoint/successful_client/invalid_quat_map_values_failure (0.00s)
        --- PASS: TestGetPositionEndpoint/successful_client/empty_quat_map_failure (0.00s)
    --- PASS: TestGetPositionEndpoint/invalid_client_no_quat_key_found_in_extra_failure (0.00s)
    --- PASS: TestGetPositionEndpoint/invalid_client_error_return_failure (0.00s)
    --- PASS: TestGetPositionEndpoint/nil_client_failure (0.00s)
=== RUN   TestGetPointCloudMapEndpoint
=== RUN   TestGetPointCloudMapEndpoint/successful_client
=== RUN   TestGetPointCloudMapEndpoint/successful_client/small_chunk_size_success
=== RUN   TestGetPointCloudMapEndpoint/successful_client/large_chunk_size_success
=== RUN   TestGetPointCloudMapEndpoint/successful_client/no_bytes_success
=== RUN   TestGetPointCloudMapEndpoint/invalid_client_error_return_failure
=== RUN   TestGetPointCloudMapEndpoint/nil_client_failure
--- PASS: TestGetPointCloudMapEndpoint (0.00s)
    --- PASS: TestGetPointCloudMapEndpoint/successful_client (0.00s)
        --- PASS: TestGetPointCloudMapEndpoint/successful_client/small_chunk_size_success (0.00s)
        --- PASS: TestGetPointCloudMapEndpoint/successful_client/large_chunk_size_success (0.00s)
        --- PASS: TestGetPointCloudMapEndpoint/successful_client/no_bytes_success (0.00s)
    --- PASS: TestGetPointCloudMapEndpoint/invalid_client_error_return_failure (0.00s)
    --- PASS: TestGetPointCloudMapEndpoint/nil_client_failure (0.00s)
=== RUN   TestGetInternalStateEndpoint
=== RUN   TestGetInternalStateEndpoint/successful_client
=== RUN   TestGetInternalStateEndpoint/successful_client/small_internal_state_chunk_size_success
=== RUN   TestGetInternalStateEndpoint/successful_client/large_internal_state_chunk_size_success
=== RUN   TestGetInternalStateEndpoint/successful_client/no_internal_state_bytes_success
=== RUN   TestGetInternalStateEndpoint/invalid_client_error_return_failure
=== RUN   TestGetInternalStateEndpoint/nil_client_failure
--- PASS: TestGetInternalStateEndpoint (0.00s)
    --- PASS: TestGetInternalStateEndpoint/successful_client (0.00s)
        --- PASS: TestGetInternalStateEndpoint/successful_client/small_internal_state_chunk_size_success (0.00s)
        --- PASS: TestGetInternalStateEndpoint/successful_client/large_internal_state_chunk_size_success (0.00s)
        --- PASS: TestGetInternalStateEndpoint/successful_client/no_internal_state_bytes_success (0.00s)
    --- PASS: TestGetInternalStateEndpoint/invalid_client_error_return_failure (0.00s)
    --- PASS: TestGetInternalStateEndpoint/nil_client_failure (0.00s)
PASS
```

File: go.mod [hunk 22]
```diff
@@ -11,12 +11,13 @@ require (
...
 	go.opencensus.io v0.24.0
-	go.viam.com/api v0.1.106
-	go.viam.com/rdk v0.2.34-0.20230407220031-b6a992c7135b
+	go.viam.com/api v0.1.109
+	go.viam.com/rdk v0.2.35-0.20230414200726-59633a8a138b
 	go.viam.com/slam v0.1.34-0.20230404174630-f82fcc184401
 	go.viam.com/test v1.1.1-0.20220913152726-5da9916c08a2
 	go.viam.com/utils v0.1.18-0.20230327140716-bfeb34d89117
 	google.golang.org/grpc v1.54.0
+	google.golang.org/protobuf v1.30.0
```
Comment (jeremyrhyde) [inline go.mod:20]: [note] added due to inject code's use of proto request/response types

File: viam-cartographer_internal_test.go (hunk 11 above)
Comment (kkufieta) [reply viam-cartographer_internal_test.go:67]: Don't have a strong opinion

File: viam-cartographer_internal_test.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (kkufieta) [reply viam-cartographer_internal_test.go]: Sounds good!

File: viam-cartographer_internal_test.go [hunk 23]
```diff
@@ -0,0 +1,392 @@
...
+	mockSLAMClient.GetPositionFunc = func(
+		ctx context.Context,
+		in *v1.GetPositionRequest,
+		opts ...grpc.CallOption,
+	) (*v1.GetPositionResponse, error) {
+		resp, err := makeGetPositionResponse(&inputPose, &inputQuat, &inputComponentRef, "not_quat")
+		return resp, err
+	}
+
+	t.Run("invalid client no quat key found in extra failure", func(t *testing.T) {
```
Comment (kkufieta) [inline viam-cartographer_internal_test.go:200]: [opt] This is what I meant we should do

```suggestion
         t.Run("invalid client no quat key found in extra failure", func(t *testing.T) {
	mockSLAMClient.GetPositionFunc = func(
		ctx context.Context,
		in *v1.GetPositionRequest,
		opts ...grpc.CallOption,
	) (*v1.GetPositionResponse, error) {
		resp, err := makeGetPositionResponse(&inputPose, &inputQuat, &inputComponentRef, "not_quat")
		return resp, err
	}
```

**Review by kkufieta — APPROVED** [review]
LGTM with opt suggestion on moving the functions into the `t.Run` in all cases
